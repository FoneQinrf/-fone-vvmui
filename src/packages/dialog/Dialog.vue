<!--
 * @Author: Fone`峰
 * @Date: 2021-04-19 16:49:56
 * @LastEditors: Fone`峰
 * @LastEditTime: 2021-05-17 10:58:52
 * @Description: file content
 * @Email: qinrifeng@163.com
 * @Github: https://github.com/FoneQinrf
-->
<template>
  <Teleport :to="teleport" :disabled="!teleport">
    <div class="vvm-dialog">
      <Popup
        @update:show="update"
        @close="close"
        v-model:show="show"
        :onClickOverlay="onClickOverlay"
        :isTeleport="true"
      >
        <div :class="classes">
          <div v-if="showTitle" class="vvm-dialog--title">
            <slot name="title">
              {{ title }}
            </slot>
          </div>
          <div class="vvm-dialog--content">
            <slot>{{ msg }}</slot>
          </div>
          <div
            v-if="showCancel || showConfirm"
            class="vvm-dialog--button vvm-hairline-top"
          >
            <Button
              v-if="showCancel"
              text
              size="large"
              @click="onCancel"
              :class="
                showCancel && showConfirm
                  ? 'vvm-hairline-right --dialog-cancel'
                  : ''
              "
            >
              {{ cancelText }}
            </Button>
            <Button
              class="--dialog-confirm"
              v-if="showConfirm"
              @click="onConfirm"
              text
              size="large"
            >
              {{ confirmText }}
            </Button>
          </div>
        </div>
      </Popup>
    </div>
  </Teleport>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref, watch } from 'vue';
import Popup from '../popup/index.vue';
import Button from '../button/index.vue';
import { forEach } from 'lodash';

const props = {
  msg: {
    type: [String, Function],
    default: ''
  },
  title: {
    type: [String, Function],
    default: '标题'
  },
  onClickOverlay: {
    type: Boolean,
    default: false
  },
  show: Boolean,
  showTitle: {
    type: Boolean,
    default: true
  },
  // showButton: {
  //   type: Boolean,
  //   default: true
  // },
  showCancel: {
    type: Boolean,
    default: false
  },
  showConfirm: {
    type: Boolean,
    default: true
  },
  cancelText: {
    type: String,
    default: '取消'
  },
  confirmText: {
    type: String,
    default: '确认'
  },
  onConfirmClose: {
    type: Boolean,
    default: false
  },
  onCancelClose: {
    type: Boolean,
    default: true
  },
  teleport: {
    type: [String, Element]
  }
};

export default defineComponent({
  name: 'Dialog',
  components: { Popup, Button },
  emits: ['update', 'close', 'cancel', 'confirm', 'update:show'],
  props,
  setup(props, { emit }) {
    const show = ref(false);

    onMounted(() => {
      show.value = props.show;
    });

    watch(
      () => props.show,
      (val) => {
        show.value = val;
      }
    );

    const update = (val: boolean) => {
      emit('update', val);
      emit('update:show', val);
    };

    const classes = computed(() => {
      return {
        [`vvm-dialog--body`]: true,
        [`--none-title`]: !props.showCancel && !props.showConfirm
      };
    });

    const close = () => {
      emit('close');
      update(false);
    };

    const onCancel = () => {
      if (props.onCancelClose) {
        show.value = false;
      }
      emit('cancel');
      update(false);
    };
    const onConfirm = () => {
      if (props.onConfirmClose) {
        show.value = false;
      }
      emit('confirm');
      update(false);
    };

    return {
      show,
      update,
      classes,
      close,
      onCancel,
      onConfirm
    };
  }
});
</script>
