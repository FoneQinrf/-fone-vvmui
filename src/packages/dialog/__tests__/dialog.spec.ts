/*
 * @Author: Fone`峰
 * @Date: 2021-05-14 10:53:16
 * @LastEditors: Fone`峰
 * @LastEditTime: 2021-05-15 21:46:20
 * @Description: file content
 * @Email: qinrifeng@163.com
 * @Github: https://github.com/FoneQinrf
 */
import { mount } from '@vue/test-utils';
import Dialog from '../Dialog.vue';
import { nextTick, createVNode } from 'vue';

describe('Dialog', () => {
  const wrapper = mount(Dialog);
  it('should component correctly', async () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('should render dialog correctly', async () => {
    wrapper.setProps({ show: true });
    await nextTick();
    expect(wrapper.find('.vvm-dialog--body').html()).toMatchSnapshot();
    wrapper.setProps({ show: false });
  });

  it('should render title text correctly', async () => {
    wrapper.setProps({ show: true, title: 'custom title' });
    await nextTick();
    expect(wrapper.find('.vvm-dialog--title').html()).toMatchSnapshot();
    expect(wrapper.find('.vvm-dialog--title').text()).toBe('custom title');
  });

  it('should render msg text correctly', async () => {
    wrapper.setProps({ show: true, msg: 'custom msg' });
    await nextTick();
    expect(wrapper.find('.vvm-dialog--content').html()).toMatchSnapshot();
    expect(wrapper.find('.vvm-dialog--content').text()).toBe('custom msg');
  });

  it('should render title slot correctly', async () => {
    const dialog = mount(Dialog, {
      props: {
        show: true
      },
      slots: {
        title: () => createVNode('span', {}, 'custom title')
      }
    });
    await nextTick();
    expect(dialog.find('.vvm-dialog--title').html()).toMatchSnapshot();
    expect(dialog.find('.vvm-dialog--title').find('span').html()).toBe(
      '<span>custom title</span>'
    );
  });

  it('should render msg slot correctly', async () => {
    const dialog = mount(Dialog, {
      props: {
        show: true
      },
      slots: {
        default: () => createVNode('span', {}, 'custom msg')
      }
    });
    await nextTick();
    expect(dialog.find('.vvm-dialog--content').html()).toMatchSnapshot();
    expect(dialog.find('.vvm-dialog--content').find('span').html()).toBe(
      '<span>custom msg</span>'
    );
  });

  it('should render cancel button and text correctly', async () => {
    const dialog = mount(Dialog, {
      props: { show: true, showCancel: true, cancelText: 'custom cancel' }
    });
    await nextTick();
    expect(dialog.find('.--dialog-cancel').html()).toMatchSnapshot();
    expect(dialog.find('.--dialog-cancel').find('.text').text()).toBe(
      'custom cancel'
    );
    dialog.setProps({ showCancel: false });
    await nextTick();
    expect(dialog.find('.--dialog-cancel').exists()).toBe(false);
  });

  it('should render confirm button and text correctly', async () => {
    expect(wrapper.find('.--dialog-confirm').html()).toMatchSnapshot();
    wrapper.setProps({ confirmText: 'custom confirm' });
    await nextTick();
    expect(wrapper.find('.--dialog-confirm').find('.text').text()).toBe(
      'custom confirm'
    );
    wrapper.setProps({ showConfirm: false });
    await nextTick();
    expect(wrapper.find('.--dialog-confirm').exists()).toBe(false);
  });

  it('should render title hidden correctly', async () => {
    wrapper.setProps({ showTitle: false });
    await nextTick();
    expect(wrapper.find('.vvm-dialog--title').exists()).toBe(false);
  });

  it('should render button hidden correctly', async () => {
    wrapper.setProps({ showConfirm: false, showCancel: false });
    await nextTick();
    expect(wrapper.find('.vvm-dialog--button').exists()).toBe(false);
  });

  it('should cancel click correctly', async () => {
    const dialog = mount(Dialog, {
      props: { show: true, showCancel: true, onCancelClose: false }
    });
    await nextTick();
    dialog.find('.--dialog-cancel').trigger('click');
    await nextTick();
    expect(dialog.vm.show).toBe(true);
    expect(dialog.find('.vvm-dialog--body').exists()).toBe(true);

    dialog.setProps({ onCancelClose: true });
    await nextTick();
    dialog.find('.--dialog-cancel').trigger('click');
    await nextTick();
    expect(dialog.vm.show).toBe(false);
    expect(dialog.find('.vvm-dialog--body').exists()).toBe(false);
  });

  it('should confirm click correctly', async () => {
    const dialog = mount(Dialog, {
      props: { show: true }
    });
    await nextTick();
    dialog.find('.--dialog-confirm').trigger('click');
    await nextTick();
    expect(dialog.vm.show).toBe(true);
    expect(dialog.find('.vvm-dialog--body').exists()).toBe(true);

    dialog.setProps({ onConfirmClose: true });
    await nextTick();
    dialog.find('.--dialog-confirm').trigger('click');
    await nextTick();
    expect(dialog.vm.show).toBe(false);
    expect(dialog.find('.vvm-dialog--body').exists()).toBe(false);
  });
});
