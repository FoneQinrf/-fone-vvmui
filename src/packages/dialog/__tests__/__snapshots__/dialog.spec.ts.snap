// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should render cancel button and text correctly 1`] = `"<button class=\\"vvm-button vvm-button--default vvm-button--text vvm-button--large vvm-hairline-right --dialog-cancel\\"><div class=\\"text\\">custom cancel</div></button>"`;

exports[`Dialog should render confirm button and text correctly 1`] = `"<button class=\\"vvm-button vvm-button--default vvm-button--text vvm-button--large --dialog-confirm\\"><div class=\\"text\\">确认</div></button>"`;

exports[`Dialog should render dialog correctly 1`] = `"<div class=\\"vvm-dialog--body --none-title\\"><div class=\\"vvm-dialog--title\\">标题</div><div class=\\"vvm-dialog--content\\"></div><div class=\\"vvm-dialog--button vvm-hairline-top\\"><!--v-if--><button class=\\"vvm-button vvm-button--default vvm-button--text vvm-button--large --dialog-confirm\\"><div class=\\"text\\">确认</div></button></div></div>"`;

exports[`Dialog should render msg slot correctly 1`] = `"<div class=\\"vvm-dialog--content\\"><span>custom msg</span></div>"`;

exports[`Dialog should render msg text correctly 1`] = `"<div class=\\"vvm-dialog--content\\">custom msg</div>"`;

exports[`Dialog should render title slot correctly 1`] = `"<div class=\\"vvm-dialog--title\\"><span>custom title</span></div>"`;

exports[`Dialog should render title text correctly 1`] = `"<div class=\\"vvm-dialog--title\\">custom title</div>"`;
